src/rust/ide/enso-protocol/src/project_manager.rs

```diff
@@ -0,0 +1,218 @@
+//! Client library for the JSON-RPC-based Project Manager service.
+
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_import_braces)]
+#![warn(unused_qualifications)]
+#![warn(unsafe_code)]
+#![warn(missing_copy_implementations)]
+#![warn(missing_debug_implementations)]
+
+use enso_prelude::*;
+
+use crate::common::UTCDateTime;
+use crate::make_rpc_method;
+use crate::make_rpc_methods;
+use crate::make_param_map;
+use crate::make_arg;
+use json_rpc::api::Result;
+use json_rpc::Handler;
+use futures::Stream;
+use serde::Serialize;
+use serde::Deserialize;
+use shapely::shared;
+use std::future::Future;
+use uuid::Uuid;
+
+
+
+// =============
+// === Event ===
+// =============
+
+/// Event emitted by the Project Manager `Client`.
+pub type Event = json_rpc::handler::Event<Notification>;
+
+
+
+// ====================
+// === Notification ===
+// ====================
+
+/// Notification generated by the Project Manager.
+#[derive(Clone,Copy,Debug,PartialEq)]
+#[derive(Serialize, Deserialize)]
+#[serde(tag="method", content="params")]
+pub enum Notification {}
+
+
+
+// ===================
+// === RPC Methods ===
+// ===================
+
+// TODO[DG]: Wrap macro_rule with #[derive(JsonRpcInterface)]
```

Comment:
Currently with such shaped macros we have no use of autocompletion of our Rust IDE plugin. That's because completion does not work with paste_item! macro.

There are two solutions for that:
1. Have macro like this:
```rust
#[derive(JsonRpcMethods)]
trait Client {
    /// Requests that the project picker open a specified project. This operation also
    /// includes spawning an instance of the language server open on the specified project.
    #[CamelCase=OpenProject,camelCase=openProject]
    fn open_project(&self, project_id:Uuid) -> IpWithSocket;

    /// Requests that the project picker close a specified project. This operation
    /// includes shutting down the language server gracefully so that it can persist state to disk as needed.
    #[CamelCase=CloseProject,camelCase=closeProject]
    fn close_project(&self, project_id:Uuid) -> ();

    /// Requests that the project picker lists the user's most recently opened
    /// projects.
    #[CamelCase=ListRecent,camelCase=listRecent]
    fn list_recent(&self, number_of_projects:u32) -> Vec<ProjectMetaData>;

    /// Requests the creation of a new project.
    #[CamelCase=CreateProject,camelCase=createProject]
    fn create_project(&self, name:String) -> Uuid;

    /// Requests the deletion of a project.
    #[CamelCase=DeleteProject,camelCase=deleteProject]
    fn delete_project(&self, project_id:Uuid) -> ();

    /// Requests a list of sample projects that are available to the user.
    #[CamelCase=ListSample,camelCase=listSample]
    fn list_sample(&self, number_of_projects:u32) -> Vec<ProjectMetaData>;
}
```
The derive macro makes RemoteClient and MockClient structures and make them implement Client trait.
Although the stuff generated by our derive macro also is not seen by completion, but here we will have at least trait - and in actual codebase we should use mostly trait, so we will be able to inject mocks instead of real clients.

2. Make `make_rpc_methods!` not using paste_item!
macro_rules! without paste_item! work well with autocompletion, so we would have all the generated stuff! The only downside is that inside macro invocation we don't have any help from compiler at all - even auto-indentation does not work!

3. A faster approach for now: leave make_rpc_methods as it is, but add trait with all the methods - and make trait without using paste_macros. This have all the downsides of 1. and 2., but still should be usable (our autocompletion still will see trait and its methods).

For me, the most elegant solution is the point 1. But point 3. is acceptable if you don't want to spend time for rewriting everything to derive macro. What do you think about it? @mwu-tow you can also put your opinion in this matter.

Reply:
Option 3 might be even preferable over 1, if we get completion for mock client methods. Client definition will be uglier but the usage will be nicer. For library code I'd prioritize usage.
We don't much care about completion in the macro invocation, as it contains only method stubs.

The primary point is anyway to get rid of paste! and shared! macros.
Paste, from what I see, is needed only to generate a structure named like `ListSampleInput`. If necessary, let's just add one more attribute to the list, like 
```
    #[CamelCase=ListSample,camelCase=listSample,CamelCaseInput=ListSampleInput]
    fn list_sample(&self, number_of_projects:u32) -> Vec<ProjectMetaData>;
```

Or we can try naming input structures `ListSample` but placing them all in the module `input`, so the usage looks like `input::ListSample`. Or something else, not sure.

Reply:
I fixed it taking Option 3. The attribute list got a bit longer, e.g.:

```
#[MethodInput=OpenProjectInput,camelCase=openProject,result=open_project_result,set_result=set_open_project_result]
```